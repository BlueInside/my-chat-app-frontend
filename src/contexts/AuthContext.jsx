import { createContext, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

const AuthContext = createContext({});

function useAuth() {
  return useContext(AuthContext);
}

const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  useEffect(() => {
    let source = axios.CancelToken.source();

    const authenticateUser = async () => {
      try {
        const url = 'http://localhost:3000/authenticate/verify-token';
        const token = localStorage.getItem('token');
        if (!token) {
          return;
        }

        const response = await axios.get(url, {
          cancelToken: source.token,
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });

        if (response.status === 200) {
          const userData = { ...response.data.user, token };
          setUser(userData);
        }
      } catch (error) {
        if (error.code === 'ERR_CANCELED') return;
        console.error('Error during validation:', error.message);
      }
    };
    authenticateUser();

    return () => {
      source.cancel('Cancelling in cleanup');
    };
  }, []);

  const login = (userData) => {
    localStorage.setItem('token', userData.token);
    setUser(userData);
  };

  const logout = () => {
    localStorage.removeItem('token');
    setUser(null);
  };

  const isAuthenticated = () => {
    return !!localStorage.getItem('token');
  };

  const value = {
    user,
    login,
    logout,
    isAuthenticated,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

AuthProvider.propTypes = {
  children: PropTypes.node,
};

export { AuthContext, AuthProvider, useAuth };
