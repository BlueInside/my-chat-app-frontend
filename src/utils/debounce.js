function debounce(fn, ms) {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn.apply(this, args);
    }, ms);
  };
}


export default debounce;
