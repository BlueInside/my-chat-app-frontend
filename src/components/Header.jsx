import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { useAuth } from '../contexts/AuthContext';

const HeaderContainer = styled.header`
  padding: 20px;
  text-align: center;
`;

const Title = styled.h1`
  color: #343a40;
  margin-bottom: 20px;
`;

const Button = styled.button`
  padding: 10px 20px;
  margin: 5px;
  background-color: #007bff;
  color: white;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  font-size: 1.1rem;

  &:hover {
    background-color: #0056b3;
  }
`;

const LogoutLink = styled.a`
  color: #007bff;
  cursor: pointer;
  margin-left: 10px;

  &:hover {
    text-decoration: underline;
  }
`;

const UserInfo = styled.p`
  margin: 10px 0;
`;

const Subtitle = styled.p`
  font-size: 1.25rem;
  margin-bottom: 20px;
`;

function Header() {
  const { user, logout } = useAuth();

  if (user) {
    return (
      <HeaderContainer>
        <Title>Welcome back, {user.username}!</Title>
        <UserInfo>Ready to jump back into the conversation?</UserInfo>
        <Link to={'/chat'}>
          <Button>Enter Chat</Button>
        </Link>
        <UserInfo>
          It&apos;s not you?
          <LogoutLink
            href="#"
            onClick={(e) => {
              e.preventDefault();
              logout();
            }}
          >
            Logout
          </LogoutLink>
        </UserInfo>
      </HeaderContainer>
    );
  }

  return (
    <HeaderContainer>
      <Title>Chat Without Boundaries</Title>
      <Subtitle>Connect instantly with people all over the world.</Subtitle>
      <Subtitle>
        Join today and start experiencing real-time communication with ease.
      </Subtitle>
      <div>
        <Link to={'/login'}>
          <Button>Login</Button>
        </Link>
        <Link to={'/register'}>
          <Button>Register</Button>
        </Link>
      </div>
    </HeaderContainer>
  );
}

export default Header;
