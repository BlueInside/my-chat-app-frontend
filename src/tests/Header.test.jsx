import { describe, it, expect, vi } from 'vitest';
import { render, screen, within } from '@testing-library/react';
import Header from '../components/Header';
import { BrowserRouter as Router } from 'react-router-dom';
import { AuthContext } from '../utils/AuthContext';

describe('Header component', () => {
  it('Checks snapshot', () => {
    const { container } = render(
      <Router>
        <Header></Header>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it('Renders the header with main, call to action and buttons to log in or register', () => {
    render(
      <Router>
        <Header></Header>
      </Router>
    );
    const header = screen.getByRole('banner');
    const loginButton = screen.getByRole('button', { name: /login/i });
    const registerButton = screen.getByRole('button', { name: /register/i });
    const actionHeading = within(header).getByRole('heading', {
      name: /Chat Without Boundaries/i,
    });

    expect(header).toBeInTheDocument();
    expect(actionHeading).toBeInTheDocument();
    expect(loginButton).toBeInTheDocument();
    expect(registerButton).toBeInTheDocument();
  });

  it(`Does not show login or register buttons when user is logged in`, () => {
    const mockUser = {
      id: '123',
      username: 'testUser',
      token: 'token',
    };

    render(
      <Router>
        <AuthContext.Provider value={{ user: mockUser, logout: vi.fn() }}>
          <Header></Header>
        </AuthContext.Provider>
      </Router>
    );
    const loginButton = screen.queryByRole('button', { name: /login/i });
    const registerButton = screen.queryByRole('button', { name: /register/i });
    const actionHeading = screen.queryByRole('heading', {
      name: /Chat Without Boundaries/i,
    });
    const userName = screen.getByText(/testUser/i);

    expect(actionHeading).not.toBeInTheDocument();
    expect(loginButton).not.toBeInTheDocument();
    expect(registerButton).not.toBeInTheDocument();
    expect(userName).toBeInTheDocument();
  });
});
