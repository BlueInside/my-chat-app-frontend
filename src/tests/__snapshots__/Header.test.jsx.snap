// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > Checks snapshot 1`] = `
<div>
  <header
    class="sc-beqWNU hsiGrh"
  >
    <h1
      class="sc-gueXAH kCxaKJ"
    >
      Chat Without Boundaries
    </h1>
    <a
      href="/login"
    >
      <button
        class="sc-dmqGRC jRohbh"
      >
        Login
      </button>
    </a>
    <a
      href="/register"
    >
      <button
        class="sc-dmqGRC jRohbh"
      >
        Register
      </button>
    </a>
  </header>
</div>
`;
